fuzz/fuzz_targets/preserve_rdata.rs

```diff
@@ -0,0 +1,227 @@
+#![no_main]
+
+//! This fuzzer checks that, if a DNS message can be decoded, it is re-encoded in a way that
+//! preserves RDATA sections of records as appropriate. Names embedded in the RDATA section of some
+//! well-known record types may be compressed or decompressed, but otherwise the RDATA should be
+//! preserved byte-for-byte. See RFC 3597, section 3:
+//!
+//! "... servers MUST also exactly preserve the RDATA of RRs of known type, except for changes due
+//! to compression or decompression where allowed by section 4 of this memo. In particular, the
+//! character case of domain names that are not subject to compression MUST be preserved."
+
+use libfuzzer_sys::fuzz_target;
+
+use hickory_proto::{
+    op::Message,
+    serialize::binary::{BinDecodable, BinEncodable},
+};
+
+fuzz_target!(|data: &[u8]| {
+    if let Ok(message) = Message::from_bytes(data) {
+        let reencoded = message.to_bytes().unwrap();
+        compare(data, &message, &reencoded);
+    }
+});
+
+fn compare(original: &[u8], message: &Message, reencoded: &[u8]) {
+    assert_eq!(original[4..6], reencoded[4..6]);
+    let query_count = u16::from_be_bytes(reencoded[4..6].try_into().unwrap());
+    assert_eq!(original[6..8], reencoded[6..8]);
+    let answer_count = u16::from_be_bytes(reencoded[6..8].try_into().unwrap());
+    assert_eq!(original[8..10], reencoded[8..10]);
+    let name_server_count = u16::from_be_bytes(reencoded[8..10].try_into().unwrap());
+    assert_eq!(original[10..12], reencoded[10..12]);
+    let additional_records_count = u16::from_be_bytes(reencoded[10..12].try_into().unwrap());
+
+    let rr_count = answer_count + name_server_count + additional_records_count;
+    let original_rrs = split_rrs(original, query_count, rr_count);
+    let reencoded_rrs = split_rrs(reencoded, query_count, rr_count);
+
+    for (original_rr, reencoded_rr) in original_rrs.into_iter().zip(reencoded_rrs.into_iter()) {
+        assert_eq!(original_rr.r#type, reencoded_rr.r#type);
+        if let Err(()) = compare_rr(original, original_rr, reencoded, reencoded_rr) {
+            println!("Parsed message: {message:?}");
+            println!("Record type: {}", original_rr.r#type);
+            println!("Original: {:02x?}", &original_rr.rdata);
+            println!("Re-encoded: {:02x?}", &reencoded_rr.rdata);
+            panic!("record RDATA was not preserved when decoding and re-encoding");
+        }
+    }
+}
+
+const RECORD_TYPE_NS: u16 = 2;
+const RECORD_TYPE_MD: u16 = 3;
+const RECORD_TYPE_MF: u16 = 4;
+const RECORD_TYPE_CNAME: u16 = 5;
+const RECORD_TYPE_SOA: u16 = 6;
+const RECORD_TYPE_MB: u16 = 7;
+const RECORD_TYPE_MG: u16 = 8;
+const RECORD_TYPE_MR: u16 = 9;
+const RECORD_TYPE_PTR: u16 = 12;
+const RECORD_TYPE_MINFO: u16 = 14;
+const RECORD_TYPE_MX: u16 = 15;
+const RECORD_TYPE_OPT: u16 = 41;
```

Comment:
Suggest spelling this as a `#[repr(u16)] enum RecordType { .. }`. (Or some other name to prevent it from crashing.)

Reply:
I'm currently using these in patterns, and changing them to enum variants would prevent that (`RecordType::NS as u16` is an expression, and thus ineligible to be part of a pattern). I could move them all into an inline module to namespace them. (unless... `#![feature(inline_const_pat)]`)